docs/testing/integration-tests.md

```diff
@@ -0,0 +1,200 @@
+ 
+## Unit Tests vs. Integration Tests
+
+Unit tests are great for ensuring that functionality works as expected at an insolated, functional-level. This might include checking that function `get_nth_fibonacci(n: u8)` works as expected, handles invalid input gracefully, etc. Unit tests in smart contracts might similarly test public functions, but can get unruly if there are several calls between accounts. As mentioned in the [unit tests](unit-tests.md) section, there is a `VMContext` object used by unit tests to mock some aspects of a transaction. One might, for instance, modify the testing context to have the `predecessor_account_id` of `"bob.near"`. The limits of unit tests become obvious with certain interactions, like transferring tokens. Since `"bob.near"` is simply a string and not an account object, there is no way to write a unit test that confirms that Alice sent Bob 6 NEAR (Ⓝ). Furthermore, there is no way to write a unit test that executes cross-contract calls. Additionally, there is no way of profiling gas usage and the execution of the call (or set of calls) on the blockchain.
+
+Integration tests provide the ability to have end-to-end testing that includes cross-contract calls, proper user accounts, access to state, structured execution outcomes, and more. In NEAR, we can make use of the `workspaces` libraries in both [Rust](https://github.com/near/workspaces-rs) and [JavaScript](https://github.com/near/workspaces-js) for this type of testing on a locally-run blockchain, testnet or mainnet.
+
+## When to Use Integration Tests
+
+You'll probably want to use integration tests when:
+
+- There are cross-contract calls.
+- There are multiple users with balance changes.
+- You'd like to gather information about gas usage and execution outcomes on-chain.
+- You want to assert the use-case execution flow of your smart contract logic works as expected.
+- You want to assert given execution patterns do not work (as expected). 
+
+## Setup
+
+Unlike unit tests in Rust (which would often live in the `src/lib.rs` file of the contract), integration tests are conventionally in a separate subdirectory of the contract, with a name of your liking (such as `tests`). Refer to this folder structure below:
+
+```sh
+├── Cargo.toml
+├── src
+│  └── lib.rs         ⟵ contract code
+├── target
+└── tests             ⟵ integration test directory
+   └── src            ⟵ optional directory for convention
+      └── tests.rs    ⟵ integration test file
+    Cargo.toml        ⟵ Cargo.toml file specifying run configuration
+```
+
+Notice we have 2 `Cargo.toml` files in this project. The first is the main `Cargo.toml` file which is used to specify the dependencies of the project, and the second which is used to specify the dependencies of the integration tests. A sample configuration for the integration tests is shown below:
```

Comment:
Is there a reason why we are describing a more complex approach with 2 `Cargo.toml` files as opposed to doing something like this?
```sh
├── Cargo.toml        ⟵ contains `dependencies` for contract and `dev-dependencies` for workspaces-rs tests
├── src
│  └── lib.rs         ⟵ contract code
├── target
└── tests             ⟵ integration test directory
   └── lib.rs         ⟵ optional directory for convention
```

Reply:
Concern separation solely. Does indeed lead to downloading deps twice. Happy to suggest your config instead.

Will adjust as to suggest your config and place an info pop up that this is also a possible route.

Reply:
Valid. What would you suggest?

Reply:
Might just be better to call these types of tests `onchain-tests` instead

Reply:
I'd prefer the structure in https://github.com/near/near-sdk-rs/tree/master/examples/fungible-token since it's pretty minimal and requires users to setup less. We wouldn't need the second Cargo.toml in that case, since we can just put test dependencies in `[dev-dependencies]`